src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs

```diff
@@ -565,10 +569,27 @@ private void PopEvalStack()
 
         public BoundNode VisitStatement(BoundNode node)
         {
-            Debug.Assert(node == null || EvalStackIsEmpty());
+#if DEBUG
+            Debug.Assert(node == null || StackDepth() == _expectedStackDepth);
+#endif
             return VisitSideEffect(node);
         }
 
+        public ImmutableArray<BoundStatement> VisitSideEffects(ImmutableArray<BoundStatement> statements)
+        {
+#if DEBUG
+            Debug.Assert(_expectedStackDepth == 0);
```

Comment:
>Debug.Assert(_expectedStackDepth == 0); [](http://example.com/codeflow?start=12&length=39)

Is this restriction really necessary? Is there something wrong with emitting side effects when ```_expectedStackDepth``` is not 0? Consider simply saving and restoring the current value.